### ISSUE TITLE
FR: Support multiple output shapes via batch_contract()

### ISSUE THREAD

**Issue by fritzo** [dgasmith/opt_einsum#65]
It would be nice to support equations like `ab,bc,cd->a,b,c,d` that produce multiple different output shapes, as e.g. is common in computing marginals of graphical models via variable elimination.
```py
a, b, c, d = 5, 6, 7, 8
x, y, z = np.random.randn(a, b), np.random.randn(b, c), np.random.randn(c, d)
result = batch_contract('ab,bc,cd->a,b,c,d', x, y, z)
assert isinstance(result, tuple) and len(result) == 4
assert result[0].shape == (a,)
assert result[1].shape == (b,)
assert result[2].shape == (c,)
assert result[3].shape == (d,)
```
Note that this is backwards incompatible with `contract`, since the result is a tuple of arrays rather than an array. A naive implementation (simplified from the [implementation in Pyro](https://github.com/uber/pyro/blob/324da9c/pyro/ops/contract.py#L567)) is
```py
def batch_contract(equation, *operands, **kwargs):
    inputs, outputs = equation.split('->')
    with shared_intermediates():
        return tuple(contract(inputs + '->' + output, *operands, **kwargs)
                     for output in outputs.split(','))
```
While the naive implementation can share some work across `output` shapes, it can miss sharing opportunities across `contract` calls because each contraction path is greedily optimized blind to the other paths.

Can we generalize the `eager` strategy to jointly optimize multiple contraction paths, possibly still in a greedy fashion?

**Comment by fritzo** [comment]
cc @eb8680

**Comment by jcmgray** [comment]
Nothing springs to mind very easily with these multiple endpoints. One thing that does, that may not be sufficient, would to implement a custom cost function favouring contractions (as in #64) already in the shared cache.

**Comment by ebatz** [comment]
> While the naive implementation can share some work across `output` shapes, it can miss sharing opportunities across `contract` calls because each contraction path is greedily optimized blind to the other paths.

An algorithm to do a joint path optimization was referenced in #9. Its implementation is pretty straightforward, but I'm unsure how to design the interface so that it doesn't blow up in complexity.

Just to mention this, your class of problem with multiple outputs is still not quite the most general; one might have various combinations of input tensors too which would permit joint optimization.

**Comment by dgasmith** [comment]
Can you expand a bit on why the interface would grow in complexity? I think the syntax proposed above is relatively straight forward, unless we want to consider optimization across communicative operations such as the CCSD diagrams linked in Hartono et al. (2006) paper.

The linked algorithms generally derive from the tensor contraction engine (TCE) project which has progressed significantly since then evolving into "TAMM" from the [NWChemEx](https://www.exascaleproject.org/getting-ready-to-model-molecular-properties-at-exascale-with-nwchemex/) project. If this is something worth exploring it would be good to reach out to NWChemEx to see if the current directions of their algorithms are worth exploring.

**Comment by ebatz** [comment]
Right, I was referring to having not only multiple outputs, but a whole list of contractions that may or may not share common subexpressions. Keeping track of all the results might become a bit of a pain then. Sorry for hijacking this thread.
